### PR TITLE
Win: add Crystal::System::DirHandle

### PR DESCRIPTION
**Pull request by txe**
Moved platform specific things from `Dir` to `Crystal::System::DirHandle`
You can find `winapi.cr` and `winerror` in this [PR #4832](https://github.com/crystal-lang/crystal/pull/4832)

### CODE REVIEW

**Comment by oprypin** [comment]
Alright, so in the other PR it is decided to keep `Errno` as is for now, so this does not need to be discussed (though it's painful to look at).

How are we feeling about efficiency of these abstractions? A class wrapping a class that ultimately just stores a pointer doesn't seem right.

File: src/dir.cr [hunk 1]
```diff
@@ -246,9 +214,7 @@ class Dir
   # Creates a new directory at the given path. The linux-style permission mode
   # can be specified, with a default of 777 (0o777).
   def self.mkdir(path, mode = 0o777)
-    if LibC.mkdir(path.check_no_null_byte, mode) == -1
-      raise Errno.new("Unable to create directory '#{path}'")
-    end
+    Crystal::System::DirHandle.mkdir(path, mode)
     0
```
Comment (bew) [inline src/dir.cr]: Not exactly related to this pr, but why does ˋmkdirˋ & ˋrmdirˋ returns ˋ0ˋ?

**Comment by txe** [comment]
I suppose, `DirHandle` cannot be a struct because it has a state. How about to make 
class Dir < Crystal::System::DirHandle?

File: src/dir.cr (hunk 1 above)
Comment (txe) [reply src/dir.cr]: I don't know, but in spec there is
`Dir.mkdir(path, 0o700).should eq(0)`

**Comment by oprypin** [comment]
No no no, a subclass cannot work, that'll leak to documentation and other such details.

You're saying it cannot be a struct (perhaps only ideologically, because technically mutable structs are fine) but I think for these internal objects we have to make an exception and think about them as an inherent part of the public objects. Make it a struct to keep efficiency at a sane level.

<strike>Though I just argued in IRC that we could easily do without `Crystal::System` namespace and keep the separation by files just as it is right now by reopening the types, then we wouldn't even need to discuss this.</strike>

**Comment by asterite** [comment]
Why can't the methods that change between operating systems be implemented directly in `Dir` without the need of introducing another type?

**Comment by oprypin** [comment]
@asterite I totally agree with you, but having the `Crystal::System` namespace was an important decision of the core team that @txe is correctly following. May be worth discussing separately?

**Comment by txe** [comment]
It was suggested to move platform specific code in System, https://github.com/crystal-lang/crystal/issues/4406. 
@RX14 proposed how to do it for Dir [here](https://github.com/crystal-lang/crystal/issues/4848)

**Comment by RX14** [comment]
Yes, `Crystal::System` looks silly here, but there are plenty of places where it doesn't look silly where it's already in use. I think we should do it this way for consistency.

File: src/dir.cr (hunk 1 above)
Comment (bcardiff) [reply src/dir.cr]: Probably they should be marked as `: Nil` for avoid leaking a useless return value

Comment (txe) [reply src/dir.cr]: I'll make another PR for it

**Comment by straight-shoota** [comment]
This now superseeded by #5447

**Comment by RX14** [comment]
#5447 was merged, closing this

**Pull request closed without merging**